PX4 integration - 2

guiccbr opened the pull request:
This Pull Request integrates PX4 specific modes and commands into mavlink_vehicles.

Currently, the flight stack (PX4 or APM-ardupilot) needs to be selected statically into mavlink_vehicles.hh, but a commit that implements the auto-detection of the flight stack will be submitted soon.

Reviewed version of #28

guiccbr committed:
PX4 AUTO, Rotate and Guided modes

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr committed:
PX4 TakeOff

PX4 does not require a sequence of commands to be sent before takeoff (aka
set_mode_guided, arm_throttle). It has instead a special mode - TAKEOFF - that
when set executes all takeoff procedures autonomously. This patch implements
set_mode(mode::TAKEOFF).

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -815,16 +862,26 @@ void mav_vehicle::request_mission_item(uint16_t item_id)
                   0, 0, request_intervals_ms::arm_disarm);
 }
 
 void mav_vehicle::takeoff()
```
I still need to take a closer look into this to make sure it will work  succesfully with PX4 and Ardupilot.

rchiossi commented on mavlink_vehicles/apm.hh:
```diff
@@ -0,0 +1,24 @@
+enum APM_CUSTOM_MODE {
+    APM_CUSTOM_MODE_GUIDED = 3,
+    APM_CUSTOM_MODE_AUTO = 4,
+    APM_CUSTOM_MODE_BRAKE =17
```
missing space

rchiossi commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -312,6 +328,7 @@ void msghandler::handle(mav_vehicle &mav, const mavlink_message_t *msg)
         return;
     }
     case MAVLINK_MSG_ID_MISSION_ACK: {
+        print_verbose("MISSION ACK RECEIVED\n");
```
Why the caps on this one?

rchiossi commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -632,8 +649,25 @@ void mav_vehicle::set_mode(mode m, int timeout)
+            break;
+        default:
+            break;
+        }
```
You don't need this empty default here. Same applies for the other switches in this patch.

rchiossi commented on mavlink_vehicles/px4.hh:
```diff
@@ -0,0 +1,50 @@
+	uint32_t data;
+	float data_float;
+};
+
```
In this file you changed from 4 to 8 spaces indent.

rchiossi commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -731,6 +735,49 @@ void mav_vehicle::set_mode(mode m, int timeout)
+        // This mode is available on PX4 only
+        if (this->flight_stack_type != firmware_type::PX4) {
+            break;
+        }
```
I think it's better to add an error message here, so it's easier to debug if someone tries to use this with apm.

rchiossi commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -815,16 +862,26 @@ void mav_vehicle::request_mission_item(uint16_t item_id)
+        break;
+    default:
+        break;
+    }
```
empty default

guiccbr replied:
I remember coverity complaining because of missing default cases, but I'm not sure. I'll take a look at it.

rchiossi replied:
Yeah, I thought that could be the case. If so, ignore my comments.

rchiossi reviewed (changes requested):
Just minor comments of this pr. Just work on those details and merge.

guiccbr commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -985,12 +988,22 @@ void mav_vehicle::takeoff()
-        send_cmd_long(MAV_CMD_NAV_TAKEOFF, 0, 0, 0, 0, 0, 0,
-                      defaults::takeoff_init_alt_m,
-                      request_intervals_ms::takeoff);
+        if (this->get_mode() != mode::GUIDED) {
```
Previously, the user of the API would have to execute all these steps before sending a takeoff command to the APM, what is not necessary for PX4.
Since we want to have APM and PX4 interchangeable transparently, I've added those commands to takeoff(). That way, the user would simply need to call takeoff() for both flight stacks. I'll submit a patch to gazebo-sitl with that change as well.

guiccbr committed:
Detect flight stack automatically

By checking 'autopilot' parameter of HEARTBEAT message.

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>

guiccbr commented on mavlink_vehicles/mavlink_vehicles.cc:
```diff
@@ -268,6 +268,19 @@ void msghandler::handle(mav_vehicle &mav, const mavlink_message_t *msg)
+        // Detect flight stack
+        switch (hb.autopilot) {
+        case MAV_AUTOPILOT_ARDUPILOTMEGA:
+            mav.flight_stack_type = firmware_type::APM;
```
I think it would be nice to change the name of the enum and the attribute to match the mavlink message as follows:

`s/mav_vehicle::flight_stack_type/mav_vehicle::autopilot/`
`s/enum class firmware_type/enum class autopilot_type/`

@rchiossi, what do you think?

rchiossi replied:
agreed. It will be easier to read the code if you are familiar with mavlink messages.

guiccbr replied:
ack

rchiossi commented on mavlink_vehicles/apm.hh:
```diff
@@ -17,8 +17,8 @@
-    APM_CUSTOM_MODE_AUTO = 4,
-    APM_CUSTOM_MODE_BRAKE =17
+    APM_CUSTOM_MODE_GUIDED = 4,
+    APM_CUSTOM_MODE_AUTO = 3,
```
Is this swap intended?

guiccbr replied:
Yes. I've swapped them unintentionally when moving those numbers to the enum. See [this](https://github.com/ArduPilot/ardupilot/blob/f0f432250060fba031258988ec91166249cb4723/ArduCopter/defines.h#L92).

rchiossi reviewed (changes requested):
LGTM. Apply the name changes, squash and merge.

guiccbr committed:
Replace flight_stack/firmware_type variable names

Replaced with autopilot_type/autopilot.

Also fixed some minor identation issues.

Signed-off-by: Guilherme Campos Camargo <guilherme.campos.camargo@intel.com>
